text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,2003 @@
+its bound should be changed to `T: ~const MetaSized` and this would not result in
+any breakage due to the previously described edition migration.
+
+```rust
+pub const fn size_of_val<T>(val: &T) -> usize
+where
+    T: ~const MetaSized,
+{
+    /* .. */
+}
+```
+
+While `MetaSized` is equivalent to the current `?Sized` bound it replaces, it
+excludes extern types (which `?Sized` by definition cannot), which prevents
+`size_of_val` from being called with extern types from [rfcs#1861][rfc_extern_types].
+Due to the changes described in [`Sized` bounds][sized-bounds] (migrating
+`T: ?Sized` to `T: const MetaSized`), changing the bound of `size_of_val` will
+not break any existing callers.
+
+These same changes apply to `align_of` and `align_of_val`.
+
+## Implementing `Copy` for runtime-sized types
+
+Runtime-sized types are value types which should be able to be used as
+locals, be copied, etc. To be able to implement [`Copy`][api_copy], these
+types would need to implement [`Clone`][api_clone], and to implement
+`Clone`, they need to implement `Sized`.
+
+This property trivially falls out of the previous proposals in this RFC:
+runtime-sized types are `Sized` (but not `const Sized`) and therefore can
+implement `Clone` and `Copy`.
+
+## Restrictions in compound types
+[restrictions-on-compound-types]: #restrictions-on-compound-types
+
+`Pointee` types cannot be used in non-`#[repr(transparent)]` compound types
+as the alignment of these types would need to be known in order to calculate field
+offsets. `const Sized` types can be used in compound types with no restrictions.
+`Sized`, `const MetaSized` and `MetaSized` types can be used in compound types, but
+only as the last element.
+
+## Compiler performance implications
+[compiler-performance-implications]: #compiler-performance-implications
+
+There is a potential performance impact within the trait system to adding
+supertraits to `Sized`, as implementation of these supertraits will need to be
+proven whenever a `Sized` obligation is being proven (and this happens very
+frequently, being a default bound). It may be necessary to implement an
+optimisation whereby `Sized`'s supertraits are assumed to be implemented and
+checking them is skipped - this should be sound as all of these traits are
+implemented by the compiler and therefore this property can be guaranteed.
+
+## Forward compatibility with supertraits implying default bound removal
+[forward-compatibility-with-supertraits-default-bound]: #forward-compatibility-with-supertraits-implying-default-bound-removal
+
+Traits which are a supertrait of any of the proposed traits will not
+automatically imply the proposed trait in any bounds where the trait is
+used, e.g.
+
+```rust
+trait NewTrait: MetaSized {}
+
+struct NewRc<T: NewTrait> {} // equiv to `T: NewTrait + Sized` as today
+```
+
+If the user wanted `T: MetaSized` then it would need to be written explicitly.
```

Comment:
Is this true in just the subtractive direction or in the additive direction as well?  I.e., this works today:

```rust
trait Tr: Sized {}
fn f<T: ?Sized + Tr>() {
    _ = const { size_of::<T>() }
}
```

Reply:
It would only apply in the subtractive direction (at least w/out the changes you're proposing in [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/410673-t-lang.2Fmeetings/topic/Design.20meeting.202025-02-05/near/499119657) and [referencing here](https://github.com/rust-lang/rfcs/pull/3729#discussion_r1951747465)). I can clarify this.

Reply:
Added to the example here which should clarify